nengo_loihi/splitter.py

```diff
@@ -4,184 +4,280 @@
-                and pre in self._chip_objects
-            ):
-                if network.config[pre].on_chip:
-                    raise BuildError(
-                        "Pre ensemble (%r) of error "
-                        "connection (%r) must not be configured "
-                        "as on_chip." % (pre, conn)
-                    )
-                self._chip_objects.remove(pre)
+        # On-chip convolutional connections are also not supported with
+        # precompute=True because the BasicSpikeGenerator cannot send population
+        # spikes correctly, meaning we have to use Snips.
+        # TODO: check with Eric if this logic is correct.
+        has_convolution = False
+        if nengo_transforms is not None:
+            has_convolution = any(
+                isinstance(conn.transform, nengo_transforms.Convolution)
+                for conn in self._conns
+            )
 
-    def _preclosure(self):  # noqa: C901
-        """Returns all objects that [in]directly send data to chip."""
-        # performs a "transitive closure" on all host objects that
-        # send data to the chip. if any of these objects receive
-        # output from the chip, then a BuildError is raised
-        precomputable = set()
+        if not has_learning and not has_convolution:
+            self._find_precomputable_objs()
+        else:
+            self._precomputable = False
+            if strict and has_learning:
+                raise BuildError(
+                    "precompute=True not supported when using learning rules"
+                )
+            elif strict and has_convolution:
+                raise BuildError(
+                    "precompute=True not supported when using convolutional connections"
+                )
 
-        # forwards and backwards adjacency lists
-        pre_to_conn = defaultdict(list)
-        post_to_conn = defaultdict(list)
+        if strict and not self._precomputable:
+            raise BuildError("Cannot precompute input, as it is dependent on output")
 
-        # data-structure for breadth-first search
+    def _find_precomputable_objs(self):
+        # Set up the breadth-first search
         queue = []
         head = 0
 
-        def mark_precomputable(obj):
-            assert isinstance(obj, (Node, Ensemble))
-            if obj not in precomputable:
-                precomputable.add(obj)
-                queue.append(obj)
-
-        # determine which connections will actually be built
-        conns = (
-            set(self.network.all_connections) | self.passthrough.to_add
-        ) - self.passthrough.to_remove
+        # Forwards and backwards adjacency lists
+        pre_to_conn = defaultdict(list)
+        post_to_conn = defaultdict(list)
 
         # Initialize queue with the pre objects on host->chip connections.
         # We assume that all `conn.pre` objects are pre-computable, and then
         # raise an error later if one of them turns out to rely on chip output.
```

Comment:
Nit: we no longer raise an error. Suggested: "raise an error" -> "set a flag to mark the host model as not precomputable".

Reply:
Right you are, thanks!